zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+impl Service<Arc<Block>> for CheckpointVerifier {
+    /// The CheckpointVerifier service has two layers of `Result`s, an inner
+    /// `checkpoint::Error`, and an outer `tokio::sync::oneshot::error::RecvError`.
+    type Response = Result<BlockHeaderHash, Error>;
+    type Error = oneshot::error::RecvError;
     type Future =
         Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
 
     fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        // We don't expect the state to exert backpressure on verifier users,
-        // so we don't need to call `state_service.poll_ready()` here.
+        // We don't expect the verifier to exert backpressure on its users.
         Poll::Ready(Ok(()))
     }
 
     fn call(&mut self, block: Arc<Block>) -> Self::Future {
-        // TODO(jlusby): Error = Report, handle errors from state_service.
-        let mut state_service = self.state_service.clone();
-        let checkpoint_list = self.checkpoint_list.clone();
-
-        async move {
-            if checkpoint_list.is_empty() {
-                return Err("the checkpoint list is empty".into());
-            };
-
-            let block_height = block
-                .coinbase_height()
-                .ok_or("the block does not have a coinbase height")?;
-
-            // TODO(teor):
-            //   - implement chaining from checkpoints to their ancestors
-            //   - should the state contain a mapping from previous_block_hash to block?
-            let checkpoint_hash = match checkpoint_list.get(&block_height) {
-                Some(&hash) => hash,
-                None => return Err("the block's height is not a checkpoint height".into()),
-            };
-
-            // Hashing is expensive, so we do it as late as possible
-            if BlockHeaderHash::from(block.as_ref()) != checkpoint_hash {
-                // The block is on a side-chain
-                return Err("the block hash does not match the checkpoint hash".into());
-            }
+        // TODO(jlusby): Error = Report
 
-            // `Tower::Buffer` requires a 1:1 relationship between `poll()`s
-            // and `call()`s, because it reserves a buffer slot in each
-            // `call()`.
-            let add_block = state_service
-                .ready_and()
-                .await?
-                .call(zebra_state::Request::AddBlock {
-                    block: block.clone(),
-                });
-
-            match add_block.await? {
-                zebra_state::Response::Added { hash } => Ok(hash),
-                _ => Err("adding block to zebra-state failed".into()),
-            }
+        // Set up a oneshot channel as the future
+        let (tx, rx) = oneshot::channel();
+
+        // Queue the block for verification, returning early on error
+        if self.insert_queued_block(block, tx).is_none() {
+            return Box::pin(rx.boxed());
```

Comment:
Would it be possible to have `insert_queued_block` construct the channel and return rx so its more obvious that there is an error being propagated in this case?

Reply:
How should `insert_queued_block` return the error status and the channel?
As a `Result<oneshot::Sender, oneshot::Sender>`?

Reply:
The doubled `Result` seems a little awkward to me.  One thing to consider is that if the check for whether there's a complete chain is fairly cheap, it may not be a big deal to check on every `call` invocation.  Then we wouldn't necessarily have to check for errors at all, since `insert_queued_block` could "error early" by firing into the channel and the outer code could check for a complete chain before returning the rx handle. This costs a little bit of work in the "happens ~never" case but could potentially simplify the logic.

Reply:
> How should `insert_queued_block` return the error status and the channel?
> As a `Result<oneshot::Sender, oneshot::Sender>`?

I think you might have Sender and Receiver mixed up but yea, that's what I was thinking. Even though it's the same object for each type in the first case the user can reasonably expect either variants of a result from a recv on the Receiver where as in the second case that Receiver should only ever return an Err variant, so it logically represents an Error coming from `insert_queued_block`.



> The doubled `Result` seems a little awkward to me.

I don't think the double result changes things here. We can still return the channel here and so long as the outer service is marked as a `Box<dyn Error>` the return type from the channel should coerce into a box. We probably need to wrap it in an `async move` though and do an `await?` internally to help the conversion happen.

Reply:
Fixed in be8a68fa4c29290bed2f616e5c2289c7db183076